File: src/main/java/no/nav/syfo/service/BeskrivelseService.java
package no.nav.syfo.service;

import no.nav.syfo.domain.Periode;
import no.nav.syfo.domain.Soknad;
import no.nav.syfo.domain.dto.SoknadPeriode;
import no.nav.syfo.domain.dto.Soknadstype;
import no.nav.syfo.domain.dto.Sporsmal;
import no.nav.syfo.domain.dto.Svar;

import java.time.LocalDate;
import java.util.Collection;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.IntStream;
import java.util.stream.Stream;

import static java.util.Arrays.asList;
import static java.util.Collections.*;
import static java.util.Optional.empty;
import static no.nav.syfo.domain.dto.Svartype.*;
import static no.nav.syfo.util.DatoUtil.norskDato;
import static no.nav.syfo.util.PeriodeMapper.jsonTilPeriode;

public class BeskrivelseService {

    public static String lagBeskrivelse(final Soknad soknad) {
        String tittel;
        switch (soknad.getSoknadstype()) {
            case ARBEIDSTAKERE:
                tittel = "Søknad om sykepenger for perioden " +
                        soknad.getFom().format(norskDato) + " - " + soknad.getTom().format(norskDato);
                break;
            case SELVSTENDIGE_OG_FRILANSERE:
                // Det kan finnes eldre søknader som mangler arbeidssituasjon
                String arbeidssituasjon = soknad.getArbeidssituasjon() != null ? soknad.getArbeidssituasjon().getNavn() : "Selvstendig Næringsdrivende / Frilanser";
                tittel = "Søknad om sykepenger fra " + arbeidssituasjon + " for perioden " +
                        soknad.getFom().format(norskDato) + " - " + soknad.getTom().format(norskDato);
                break;
            case OPPHOLD_UTLAND:
                tittel = "Søknad om å beholde sykepenger i utlandet";
                break;
            default:
                throw new RuntimeException("Beskrivelse er ikke implementert for søknadstype: " + soknad.getSoknadstype());
        }
        return tittel + (soknad.getKorrigerer() != null ? " KORRIGERING" : "") + "\n" +
                beskrivArbeidsgiver(soknad) +
                Optional.ofNullable(soknad.getSoknadPerioder()).map(BeskrivelseService::beskrivPerioder).orElse("") +
                soknad.getSporsmal().stream()
                        .filter(BeskrivelseService::sporsmalSkalVises)
                        .map(sporsmal -> beskrivSporsmal(sporsmal, 0))
                        .filter(Optional::isPresent)
                        .map(Optional::get)
                        .collect(Collectors.joining("\n"));
    }

    private static String beskrivPerioder(final List<SoknadPeriode> perioder) {
        return IntStream.range(0, perioder.size())
                .mapToObj(i -> {
                    SoknadPeriode soknadPeriode = perioder.get(i);
                    String faktiskGrad = soknadPeriode.getFaktiskGrad() != null
                            ? "Oppgitt faktisk arbeidsgrad: " + soknadPeriode.getFaktiskGrad() + "\n"
                            : "";
                    return "\nPeriode " + (i + 1) + ":\n" +
                            soknadPeriode.getFom().format(norskDato) + " - " + soknadPeriode.getTom().format(norskDato) + "\n" +
                            "Grad: " + soknadPeriode.getGrad() + "\n" +
                            faktiskGrad;
                })
                .collect(Collectors.joining());

    }

    private static String beskrivArbeidsgiver(final Soknad soknad) {
        return soknad.getSoknadstype() == Soknadstype.ARBEIDSTAKERE
            ?  "\nArbeidsgiver: " + soknad.getArbeidsgiver() + "\n"
            : "";
    }

    private static boolean sporsmalSkalVises(final Sporsmal sporsmal) {
        switch (sporsmal.getTag()) {
            case "ANSVARSERKLARING":
            case "BEKREFT_OPPLYSNINGER":
                return false;
            case "ARBEIDSGIVER":
                return true;
            default:
                return !"NEI".equals(getForsteSvarverdi(sporsmal));
        }
    }

    private static Optional<String> beskrivSporsmal(final Sporsmal sporsmal, final int dybde) {
        final String innrykk = "\n" + String.join("", nCopies(dybde, "    "));
        final List<String> svarverdier = getSvarverdier(sporsmal);
        return svarverdier.isEmpty() && Stream.of(CHECKBOX_GRUPPE, RADIO_GRUPPE, RADIO_GRUPPE_TIMER_PROSENT).noneMatch(svartype -> svartype.equals(sporsmal.getSvartype()))
                ? empty()
                : Optional.of(formatterSporsmalOgSvar(sporsmal).stream()
                .map(sporsmalOgSvar -> innrykk + sporsmalOgSvar)
                .collect(Collectors.joining()) +
                getUndersporsmalIgnorerRadioIGruppeTimerProsent(sporsmal).stream()
                        .map(undersporsmal -> beskrivSporsmal(undersporsmal, getNesteDybde(sporsmal, dybde)))
                        .filter(Optional::isPresent)
                        .map(Optional::get)
                        .collect(Collectors.joining("\n")));
    }

    private static int getNesteDybde(final Sporsmal sporsmal, final int dybde) {
        return Stream.of(RADIO_GRUPPE, RADIO_GRUPPE_TIMER_PROSENT).anyMatch(svartype -> svartype.equals(sporsmal.getSvartype()))
                ? dybde
                : dybde + 1;
    }

    private static List<Sporsmal> getUndersporsmalIgnorerRadioIGruppeTimerProsent(final Sporsmal sporsmal) {
        return RADIO_GRUPPE_TIMER_PROSENT == sporsmal.getSvartype()
                ? sporsmal.getUndersporsmal().stream()
                .map(Sporsmal::getUndersporsmal)
                .flatMap(Collection::stream)
                .collect(Collectors.toList())
                : sporsmal.getUndersporsmal();
    }

    private static List<String> formatterSporsmalOgSvar(final Sporsmal sporsmal) {
        switch (sporsmal.getSvartype()) {
            case CHECKBOX:
            case CHECKBOX_GRUPPE:
            case RADIO:
            case RADIO_GRUPPE:
            case RADIO_GRUPPE_TIMER_PROSENT:
                return singletonList(sporsmal.getSporsmalstekst());
            case JA_NEI:
                return asList(sporsmal.getSporsmalstekst(), "JA".equals(getForsteSvarverdi(sporsmal)) ? "Ja" : "Nei");
            case DATO:
                return asList(sporsmal.getSporsmalstekst(), formatterDato(getForsteSvarverdi(sporsmal)));
            case PERIODE:
                return asList(sporsmal.getSporsmalstekst(), formatterPeriode(getForsteSvarverdi(sporsmal)));
            case PERIODER:
                return Stream.concat(Stream.of(sporsmal.getSporsmalstekst()),
                        getSvarverdier(sporsmal).stream()
                                .map(BeskrivelseService::formatterPeriode))
                        .filter(Objects::nonNull)
                        .collect(Collectors.toList());
            case LAND:
                return Stream.concat(Stream.of(sporsmal.getSporsmalstekst()),
                        getSvarverdier(sporsmal).stream()
                                .map(BeskrivelseService::formatterLand))
                        .filter(Objects::nonNull)
                        .collect(Collectors.toList());
            case TALL:
                return Stream.of(sporsmal.getSporsmalstekst(),
                        getForsteSvarverdi(sporsmal) + " " + sporsmal.getUndertekst())
                        .filter(Objects::nonNull)
                        .collect(Collectors.toList());
            case TIMER:
                return asList(sporsmal.getSporsmalstekst(), getForsteSvarverdi(sporsmal) + " timer");
            case PROSENT:
                return asList(sporsmal.getSporsmalstekst(), getForsteSvarverdi(sporsmal) + " prosent");
            case FRITEKST:
                return asList(sporsmal.getSporsmalstekst(), getForsteSvarverdi(sporsmal));
            default:
                return emptyList();
        }
    }

    private static List<String> getSvarverdier(final Sporsmal sporsmal) {
        return sporsmal.getSvar().stream()
                .map(Svar::getVerdi)
                .collect(Collectors.toList());
    }

    private static String getForsteSvarverdi(Sporsmal sporsmal) {
        return sporsmal.getSvar().isEmpty()
                ? ""
                : sporsmal.getSvar().get(0).getVerdi();
    }

    private static String formatterDato(final String svarverdi) {
        return LocalDate.parse(svarverdi).format(norskDato);
    }

    private static String formatterPeriode(final String svarverdi) {
        final Periode periode = jsonTilPeriode(svarverdi);
        return periode.getFom().format(norskDato) + " - " +
                periode.getTom().format(norskDato);
    }

    private static String formatterLand(final String svarverdi) {
        return "- " + svarverdi;
    }
}

File: src/main/kotlin/no/nav/syfo/service/BeskrivelseService.kt
package no.nav.syfo.service

import no.nav.syfo.domain.Soknad
import no.nav.syfo.domain.dto.Arbeidssituasjon.ARBEIDSTAKER
import no.nav.syfo.domain.dto.Avsendertype.SYSTEM
import no.nav.syfo.domain.dto.Merknad
import no.nav.syfo.domain.dto.Soknadstype.*
import no.nav.syfo.domain.dto.Sporsmal
import no.nav.syfo.domain.dto.Svartype.*
import no.nav.syfo.util.DatoUtil.norskDato
import no.nav.syfo.util.PeriodeMapper.jsonTilPeriode
import org.slf4j.LoggerFactory
import java.time.LocalDate
import java.util.*
import java.util.Collections.nCopies

val log = LoggerFactory.getLogger("no.nav.syfo.service.BeskrivelseService")

fun lagBeskrivelse(soknad: Soknad): String {
    return soknad.meldingDersomEgenmeldtSykmelding() +
        soknad.meldingDersomAvsendertypeErSystem() +
        soknad.lagTittel() +
        soknad.erKorrigert() + "\n" +
        soknad.beskrivMerknaderFraSykmelding() +
        soknad.beskrivKvitteringer() +
        soknad.beskrivArbeidsgiver() +
        soknad.beskrivPerioder() +
        soknad.beskrivFaktiskGradFrilansere() +
        soknad.sporsmal
            .filter { it.skalVises() }
            .map { sporsmal -> beskrivSporsmal(sporsmal, 0) }
            .filter { it.isNotBlank() }
            .joinToString("\n")
}

private fun Soknad.beskrivMerknaderFraSykmelding(): String {
    return if (merknaderFraSykmelding?.isNotEmpty() == true) {
        merknaderFraSykmelding
            .joinToString(separator = "\n", postfix = "\n") { it.beskrivMerknad() }
    } else ""
}

private fun Soknad.beskrivKvitteringer(): String {
    return if (kvitteringer != null && kvitteringer.isNotEmpty()) {
        "\nSøknaden har vedlagt ${kvitteringer.size} kvitteringer med en sum på ${kvitteringSum.toString().formatterBelop()} kr\n"
    } else ""
}

private fun Merknad.beskrivMerknad(): String {
    return when (type) {
        "UGYLDIG_TILBAKEDATERING" -> "OBS! Sykmeldingen er avslått grunnet ugyldig tilbakedatering"
        "TILBAKEDATERING_KREVER_FLERE_OPPLYSNINGER" -> "OBS! Tilbakedatert sykmelding er til vurdering"
        "TILBAKEDATERT_PAPIRSYKMELDING" -> "OBS! Sykmeldingen sendt til NAY for manuell kontroll grunnet tilbakedatering"
        else -> {
            log.warn("Ukjent merknadstype $type")
            "OBS! Sykmeldingen har en merknad $this"
        }
    }
}

private fun Soknad.meldingDersomEgenmeldtSykmelding() =
    if (egenmeldtSykmelding == true)
        "Denne søknaden hører til en egenmeldt sykmelding\n"
    else
        ""

private fun Soknad.meldingDersomAvsendertypeErSystem() =
    if (avsendertype == SYSTEM)
        "Denne søknaden er autogenerert på grunn av et registrert dødsfall\n"
    else
        ""

private fun Soknad.lagTittel() =
    when (soknadstype) {
        ARBEIDSTAKERE -> "Søknad om sykepenger for perioden ${fom!!.format(norskDato)} - ${tom!!.format(norskDato)}"
        SELVSTENDIGE_OG_FRILANSERE -> {
            // Det kan finnes eldre søknader som mangler arbeidssituasjon
            val arbeidssituasjon = arbeidssituasjon?.navn ?: "Selvstendig Næringsdrivende / Frilanser"
            "Søknad om sykepenger fra $arbeidssituasjon for perioden ${fom!!.format(norskDato)} - ${
            tom!!.format(
                norskDato
            )
            }"
        }
        OPPHOLD_UTLAND -> "Søknad om å beholde sykepenger utenfor EØS"
        ARBEIDSLEDIG -> "Søknad om sykepenger for arbeidsledig"
        BEHANDLINGSDAGER -> "Søknad med enkeltstående behandlingsdager"
        ANNET_ARBEIDSFORHOLD -> "Søknad om sykepenger med uavklart arbeidssituasjon"
        REISETILSKUDD -> "Søknad om reisetilskudd for perioden ${fom!!.format(norskDato)} - ${tom!!.format(norskDato)}"
        GRADERT_REISETILSKUDD -> "Søknad om sykepenger med reisetilskudd for perioden ${fom!!.format(norskDato)} - ${tom!!.format(norskDato)}"
    }

private fun Soknad.erKorrigert() =
    korrigerer?.let { " KORRIGERING" } ?: ""

private fun Soknad.beskrivArbeidsgiver() =
    if (arbeidssituasjon === ARBEIDSTAKER)
        "\nArbeidsgiver: $arbeidsgiver" +
            "\nOrganisasjonsnummer: $orgNummer\n"
    else
        ""

private fun Soknad.beskrivPerioder() =
    soknadPerioder?.mapIndexed { index, periode ->
        "\nPeriode ${index + 1}:\n" +
            "${periode.fom!!.format(norskDato)} - ${periode.tom!!.format(norskDato)}\n" +
            (
                periode.grad?.let { grad ->
                    if (soknadstype != REISETILSKUDD)
                        "Grad: ${grad}\n"
                    else
                        ""
                } ?: ""
                ) +
            (
                periode.faktiskGrad?.let { faktiskGrad ->
                    "Oppgitt faktisk arbeidsgrad: $faktiskGrad\n"
                } ?: ""
                )
    }?.joinToString("") ?: ""

// Er egen beskrivelse for selvstendig/frilanser fordi SoknadDTO Periode ikke inneholder feltet faktiskGrad før Mars 2020, slik som arbeidstakere
private fun Soknad.beskrivFaktiskGradFrilansere(): String {
    if (soknadstype === SELVSTENDIGE_OG_FRILANSERE) {
        val harJobbetMerEnnGradert = sporsmal
            .asSequence()
            .filter { it.tag.startsWith("JOBBET_DU_GRADERT_") || it.tag.startsWith("JOBBET_DU_100_PROSENT_") }
            .any { it.svar?.asSequence()?.any { svar -> svar.verdi == "JA" } ?: false }

        val periodeMedFaktiskGrad = soknadPerioder?.any { it.faktiskGrad != null } ?: false

        if (harJobbetMerEnnGradert && !periodeMedFaktiskGrad) {
            return """
                
                OBS! Brukeren har jobbet mer enn uføregraden i sykmeldingen.
                Se oppgitt arbeidsgrad lengre ned i oppgaven
                
            """.trimIndent()
        }
    }
    return ""
}

private fun Sporsmal.skalVises() =
    when (tag) {
        "ANSVARSERKLARING", "BEKREFT_OPPLYSNINGER", "EGENMELDINGER", "FRAVER_FOR_BEHANDLING" -> false
        "ARBEIDSGIVER" -> true
        "UTBETALING" -> true
        "FRISKMELDT" -> "NEI" == forsteSvarverdi()
        else -> "NEI" != forsteSvarverdi()
    }

private fun beskrivSporsmal(sporsmal: Sporsmal, dybde: Int): String {
    val innrykk = "\n" + nCopies(dybde, "    ").joinToString("")
    val svarverdier = sporsmal.svarverdier()

    return if (svarverdier.isEmpty() && sporsmal.svartype !in listOf(
            CHECKBOX_GRUPPE,
            RADIO_GRUPPE,
            RADIO_GRUPPE_TIMER_PROSENT,
            INFO_BEHANDLINGSDAGER
        )
    ) {
        ""
    } else {
        sporsmal.formatterSporsmalOgSvar().joinToString("") { sporsmalOgSvar ->
            innrykk + sporsmalOgSvar
        }.plus(
            sporsmal.undersporsmalIgnorerRadioIGruppeTimerProsent()
                ?.map { beskrivSporsmal(it, getNesteDybde(sporsmal, dybde)) }
                ?.filter { it.isNotBlank() }
                ?.joinToString("\n")
                ?: ""
        )
    }
}

private fun getNesteDybde(sporsmal: Sporsmal, dybde: Int): Int {
    return when (sporsmal.svartype) {
        RADIO_GRUPPE, RADIO_GRUPPE_TIMER_PROSENT -> dybde
        else -> dybde + 1
    }
}

private fun Sporsmal.undersporsmalIgnorerRadioIGruppeTimerProsent(): List<Sporsmal>? {
    return if (svartype === RADIO_GRUPPE_TIMER_PROSENT)
        undersporsmal!!
            .filter { it.kriterieForVisningAvUndersporsmal?.name == it.forsteSvarverdi() }
            .map { it.undersporsmal }
            .flatMap { it!! }
            .toList()
    else
        undersporsmal
}

private fun Sporsmal.formatterSporsmalOgSvar(): List<String> {
    return when (svartype) {
        CHECKBOX, CHECKBOX_GRUPPE, RADIO, RADIO_GRUPPE, RADIO_GRUPPE_TIMER_PROSENT, INFO_BEHANDLINGSDAGER ->
            listOfNotNull(sporsmalstekst)
        IKKE_RELEVANT -> emptyList()
        JA_NEI -> listOfNotNull(sporsmalstekst, if ("JA" == forsteSvarverdi()) "Ja" else "Nei")
        DATO -> listOfNotNull(sporsmalstekst, formatterDato(forsteSvarverdi()))
        PERIODE -> listOfNotNull(sporsmalstekst, formatterPeriode(forsteSvarverdi()))
        PERIODER -> listOfNotNull(sporsmalstekst) + svarverdier().map { formatterPeriode(it) }
        LAND -> listOfNotNull(sporsmalstekst) + svarverdier().map { formatterLand(it) }
        TALL, KILOMETER -> listOfNotNull(sporsmalstekst, forsteSvarverdi() + " " + undertekst)
        BELOP -> listOfNotNull(sporsmalstekst, forsteSvarverdi().formatterBelop() + " " + undertekst)
        TIMER -> listOfNotNull(sporsmalstekst, forsteSvarverdi() + " timer")
        PROSENT -> listOfNotNull(sporsmalstekst, forsteSvarverdi() + " prosent")
        FRITEKST -> listOfNotNull(sporsmalstekst, forsteSvarverdi())
        RADIO_GRUPPE_UKEKALENDER -> listOfNotNull(formatterBehandlingsdato(forsteSvarverdi()))
        DATOER -> listOfNotNull(sporsmalstekst) + svarverdier().map { formatterDato(it) }
        else -> emptyList()
    }
}

private fun Sporsmal.svarverdier(): List<String> {
    return svar?.mapNotNull { it.verdi } ?: emptyList()
}

private fun Sporsmal.forsteSvarverdi(): String {
    return svar?.firstOrNull()?.verdi ?: ""
}

private fun String.formatterBelop(): String {
    val context = this.toInt()
    val kr = context / 100
    val øre = context % 100
    return "%,d,%02d".format(locale = Locale("nb"), kr, øre)
}

private fun formatterDato(svarverdi: String?): String {
    return LocalDate.parse(svarverdi!!).format(norskDato)
}

private fun formatterPeriode(svarverdi: String?): String {
    val (fom, tom) = jsonTilPeriode(svarverdi)
    return fom.format(norskDato) + " - " +
        tom.format(norskDato)
}

private fun formatterLand(svarverdi: String): String {
    return "- $svarverdi"
}

private fun formatterBehandlingsdato(svarverdi: String): String {
    return try {
        formatterDato(svarverdi)
    } catch (e: Exception) {
        "Ikke til behandling"
    }
}
